prov/util/src/util_mr_monitor.c

```diff
@@ -0,0 +1,151 @@
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <ofi_mr.h>
+
+static struct ofi_mr_monitor global_mr_monitor;
+
+static inline int
+util_check_mr_monitor_attr_sanity(struct ofi_mr_monitor_attr *attr)
+{
+	/* callbacks and provider must be provided */
+	if (!attr || !attr->prov ||
+	    !attr->init || !attr->open || !attr->close ||
+	    !attr->subscribe || !attr->unsubscribe || !attr->get_event)
+		return -FI_EINVAL;
+
+	/* valid otherwise */
+	return FI_SUCCESS;
+}
+
+int ofi_mr_monitor_init(struct ofi_mr_monitor_attr *attr)
+{
+	int ret;
+
+	ret = util_check_mr_monitor_attr_sanity(attr);
+	if (ret)
+		return ret;
+
+	global_mr_monitor.attr = *attr;
+	fastlock_init(&global_mr_monitor.lock);
+	global_mr_monitor.ref_cnt = 0;
+	dlist_init(&global_mr_monitor.rq_head);
+
+	return global_mr_monitor.attr.init();
+}
+
+int ofi_mr_monitor_open(struct ofi_mr_monitor **mr_monitor)
+{
+	int ret = FI_SUCCESS;
+
+	fastlock_acquire(&global_mr_monitor.lock);
+	if (global_mr_monitor.ref_cnt == 0)
+		ret = global_mr_monitor.attr.open(&global_mr_monitor.notifier);
+
+	if (!ret)
+		global_mr_monitor.ref_cnt++;
+	fastlock_release(&global_mr_monitor.lock);
+
+	if (!ret)
+		*mr_monitor = &global_mr_monitor;
+
+	return ret;
+}
+
+int ofi_mr_monitor_close(struct ofi_mr_monitor *mr_monitor)
+{
+	int ret = FI_SUCCESS;
+
+	fastlock_acquire(&mr_monitor->lock);
+	if (mr_monitor->ref_cnt == 0)
+		ret = -FI_EINVAL;
+
+	if (mr_monitor->ref_cnt == 1)
+		ret = mr_monitor->attr.close(mr_monitor->notifier);
+
+	if (!ret)
+		mr_monitor->ref_cnt--;
+	fastlock_release(&mr_monitor->lock);
+
+	return ret;
+}
+
+static void util_mr_monitor_read_events(struct ofi_mr_monitor *mr_monitor)
+{
+	struct ofi_mr_monitor_context *context = NULL;
+	struct ofi_mr_monitor_rq *rq;
+	int ret, len = (int)sizeof(uint64_t);
+
+	do {
+		ret = mr_monitor->attr.get_event(mr_monitor->notifier,
+						 (void *)&context, len);
+		if (ret != len) {
+			FI_DBG(UTIL_MR_MONITOR_GET_PROV(mr_monitor), FI_LOG_MR,
+			       "no more events to be read\n");
+			break;
+		}
+
+		FI_DBG(UTIL_MR_MONITOR_GET_PROV(mr_monitor), FI_LOG_MR,
+		       "found event, context=%p rq=%p cookie=%lx\n",
+		       context, context->rq, context->cookie);
+
+		rq = context->rq;
+		fastlock_acquire(&rq->lock);
+		dlist_insert_tail(&context->entry, &rq->list);
+		fastlock_release(&rq->lock);
```

Comment:
The same subscription could have multiple events reported for it.  We need to protect against it being inserted onto the list twice, resulting in a corruption.  I would add a check
```
if (subscript->entry.next == subscript->entry.prev) /* or dlist_empty(&subscript->entry) */
    dlist_insert_tail(...);
```

Reply:
@shefty  Nice point!
Since we subscribes only on `UNMAPPED` event, we must process only first, the others should be ignored.


But I guess, it will be more natural if we will add support of both `MAPPED`/`UNMAPPED` events. So, this check will be unnecessary, because we will provide to user all events. User takes appropriate actions for each reported event.

Reply:
> But I guess, it will be more natural if we will add support of both MAPPED/UNMAPPED events.

So please give me some motivation here. Why track MAPPED events? If a notifier supports this, and there is no guarantee that a notifier does, what does this provide to the providers who use it? Is it used by a provider to update a memory registration ?

> The same subscription could have multiple events reported for it.

Is this a property of the implementation of the memory notification system that verbs uses today? This isn't true for the implementation that the GNI provider uses today.